src/work/WorkScheduler.h

```diff
@@ -47,7 +47,7 @@ class WorkScheduler : public Work
     std::shared_ptr<T>
     scheduleWork(Args&&... args)
     {
-        if (isAborting())
+        if (isAborting() || isDone())
```

Comment:
I agree that we should do this. I also think that for this specific issue it is safer if we do not trigger `scheduleSelfCheck` at all during shutdown. Right now we still execute the code in `scheduleSelfCheck`, so this creates a possibility that we invoke functions on objects that are already shutdown, which would be illegal. 

So I think we also need to cancel mSelfCheckTimer in ApplicationImpl::gracefulStop

Reply:
Makes sense! Updated it.

```
^C2021-08-09T17:22:04.913 GBAMW [default INFO] got signal 2, shutting down
2021-08-09T17:22:04.913 GBAMW [Work TRACE] Shutting down: wait-for-lcl
2021-08-09T17:22:04.913 GBAMW [Work DEBUG] work wait-for-lcl : WORK_WAITING -> WORK_ABORTING
2021-08-09T17:22:04.913 GBAMW [Work TRACE] Shutting down: self-check
2021-08-09T17:22:04.913 GBAMW [Work DEBUG] work self-check : WORK_WAITING -> WORK_ABORTING
2021-08-09T17:22:04.913 GBAMW [Work TRACE] Shutting down: work-scheduler
2021-08-09T17:22:04.913 GBAMW [Work DEBUG] work work-scheduler : WORK_WAITING -> WORK_ABORTING
2021-08-09T17:22:04.913 GBAMW [Work TRACE] Abort progress for wait-for-lcl: done
2021-08-09T17:22:04.913 GBAMW [Work TRACE] resetting wait-for-lcl
2021-08-09T17:22:04.913 GBAMW [Work DEBUG] work wait-for-lcl : WORK_ABORTING -> WORK_ABORTED
2021-08-09T17:22:04.913 GBAMW [Work TRACE] Abort progress for self-check: still aborting
2021-08-09T17:22:04.913 GBAMW [Work TRACE] Abort progress for work-scheduler: still aborting
2021-08-09T17:22:04.913 GBAMW [Work TRACE] work-scheduler: waiting for children to abort.
2021-08-09T17:22:04.913 GBAMW [Work TRACE] Abort progress for work-scheduler: still aborting
2021-08-09T17:22:04.913 GBAMW [Work TRACE] Abort progress for self-check: done
2021-08-09T17:22:04.913 GBAMW [Work TRACE] resetting self-check
2021-08-09T17:22:04.913 GBAMW [Work DEBUG] work self-check : WORK_ABORTING -> WORK_ABORTED <---- self-check aborted
2021-08-09T17:22:04.913 GBAMW [Work TRACE] Abort progress for work-scheduler: still aborting
2021-08-09T17:22:04.913 GBAMW [Work TRACE] work-scheduler: waiting for children to abort.
2021-08-09T17:22:04.913 GBAMW [Work TRACE] Abort progress for work-scheduler: done
2021-08-09T17:22:04.913 GBAMW [Work TRACE] resetting work-scheduler
2021-08-09T17:22:04.913 GBAMW [Work DEBUG] work work-scheduler : WORK_ABORTING -> WORK_ABORTED
2021-08-09T17:22:04.913 GBAMW [Bucket TRACE] b72bc26d469a3f5079510767e2ea4773e02da574dec72e2f32f5828b79e0af40 referenced by bucket list
2021-08-09T17:22:04.913 GBAMW [Bucket TRACE] 0000000000000000000000000000000000000000000000000000000000000000 referenced by bucket list
2021-08-09T17:22:05.914 GBAMW [default INFO] Application destructing
2021-08-09T17:22:05.914 GBAMW [Bucket WARNING] BucketManager: shutdown is called more than once
2021-08-09T17:22:05.914 GBAMW [default DEBUG] Joining 11 worker threads
2021-08-09T17:22:05.914 GBAMW [default DEBUG] Joined all 11 threads
2021-08-09T17:22:05.914 GBAMW [default INFO] Application destroyed
2021-08-09T17:22:05.915 GBAMW [default DEBUG] TmpDir deleted: buckets/tmp/process-ed4282c355736d14
2021-08-09T17:22:05.915 GBAMW [default DEBUG] TmpDir deleted: buckets/tmp/history-82fdf5d4107e39e6
2021-08-09T17:22:05.915 GBAMW [default DEBUG] TmpDirManager cleaning: buckets/tmp
```

After `self-check` is aborted above, the log indicates that it didn't even try to schedule anything. As a comparison, without this PR, it would print something like

```
2021-08-06T15:14:12.909 GDH22 [Work TRACE] Starting self-check
2021-08-06T15:14:12.909 GDH22 [Work TRACE] resetting self-check
2021-08-06T15:14:12.909 GDH22 [Work DEBUG] work self-check : WORK_PENDING -> WORK_RUNNING 
```
after it has been aborted.